### ISSUE TITLE
runblocking causes unintuitive deadlock with specified dispatcher

### ISSUE THREAD

**Issue by Groostav** [Kotlin/kotlinx.coroutines#1397]
in short: `runBlocking(JavaFx) { runBlocking { }}` runs where `runBlocking(JavaFx) { runBlocking(JavaFx) { }}` causes deadlock.

This is likely a follow-up to #202

All three of these tests should pass, but note the one with `@Ignore` does not

```kt
    @Ignore("hangs")
    @Test fun `when attempting to use nested runblocking with JavaFx should hang`() {
        println("running on thread ${Thread.currentThread()}")

        runBlocking(Dispatchers.JavaFx){
            println("running in runBlocking() on thread ${Thread.currentThread()}")

            runBlocking(Dispatchers.JavaFx){
                println("running in nested runblocking() on thread ${Thread.currentThread()}")
            }
        }
    }

    @Test fun `when attemping to use nested runblocking without javafx should not hang`() {
        runBlocking {
            runBlocking {
                println("hi mom!")
            }
        }
    }

    @Test fun `when attempting to use runblocking on javafx thread should not hang`() {

        runBlocking(Dispatchers.JavaFx) {
            runBlocking {
                println("hi mom! from ${Thread.currentThread()}")
            }
        }
    }
```

**Comment by fvasco** [comment]
In the official documentation:

> When CoroutineDispatcher is explicitly specified in the context, then the new coroutine runs in the context of the specified dispatcher while the current thread is blocked.

So this is a _feature_.

**Comment by Groostav** [comment]
Is there/should there be a way for me to specify that "I want to use dispatcher X if none is specified, else you should use the parent event loops dispatcher if it is using dispatcher X"?

This seems like a critically important feature to me WRT UI threading on big code-bases. A reentrant runBlocking(JavaFx) would make my projects adoption of coroutines a great deal easier.

**Comment by elizarov** [comment]
@Groostav Can you, please, add more details on a specific use-case where you run into this problem when writing UI code. Why would you be running into a case of having to use `runBlocking` inside another `runBlocking` call on the same dispatcher? How does it happen?

**Comment by Groostav** [comment]
We use guava events on view-binding code fairly regularly, so a standard UI element might be implemented like this:

```kt
@ViewBinding lateinit var someViewElement: Label

@Subscribe fun onSomeModelEvent(event: ModelEvent) = runBlocking(Dispatchers.Main) {
  someViewElement.data = event.data
}


fun complexUpstackBusinessLogic() {
  //....
  runBlocking(Dispatchers.Main){
    //...
    eventBus.post(ModelEvent())
  }
}
```

in the case of guava's event bus, that event may be posted up-stack by the call `eventBus.post(ModelEvent)())` and is itself tied to complciated business logic that may itself employ `runBlocking` blocks. 

But I would propose that this generalizes: why support nested event loops at all? A reentrant `runBlocking` is really nice because it means you can move synchronously into a coroutine regardless of context without worrying about deadlock. If you have to deal with java code above you on your callstack (eg guava event bus, a java-style cache, the java streams API, etc etc), with perhaps your business logic using coroutines being above that java code on the callstack, then a nested-event-loop style `runBlocking` represents a really great tool to kind of re-attach to your coroutines.

**Comment by Groostav** [comment]
oh, the event loop is keyed on `ContinuationInterceptor` which is the same key as the dispatcher. So my suggestion would involve a rather large API change.

```kt
fun runBlocking(){
    val currentThread = Thread.currentThread()
    val contextInterceptor = context[ContinuationInterceptor]
    val parentEventLoop = context[EventLoop]
    val privateEventLoop = contextInterceptor == null || (parentEventLoop?.dispatcher != contextInterceptor)

val newContext = GlobalScope.newCoroutineContext(
    if(privateEventLoop) context + eventLoop + eventLoop.asCoroutineDispatcher() else context
)
```

and you would have to modify `interface EventLoop` and `abstract class EventLoopBase`  to be cleverly keyed by 2 things. Hmm.

**Comment by elizarov** [comment]
@Groostav Can you, please, why do you need to use `runBlocking` in your `onXxxEvent` functions? Can you just replaces them with `launch`, e.g.:

```
@Subscribe fun onSomeModelEvent(event: ModelEvent) = mainScope.launch {
  someViewElement.data = event.data
}
```

That is how the coroutines were intended to be use with synchronous events and it is actually more performant than using `runBlocking` (has less infrastructure overhead). Would that solve your `runBlocking` issue?

**Comment by qwwdfsad** [comment]
>Regarding "its a feature", I cant imagine anybody is shipping software with the expectation that runBlocking(Main){ runBlocking (Main) deadlocks, so cant a simple fix be to check for this case?

This can be done, though it changes the behavior of nested `runBlocking` for a multi-threaded dispatcher. So let's focus on your domain problem first and then decide whether this solution is the most appropriate.

I wonder why `launch` is not sufficient in your use case. To save an additional dispatch, `launch(Dispatchers.JavaFx.immediate)` can be used as well

**Comment by Groostav** [comment]
I don't understand: the suggestion is to use an asynchronous device (ie `launch`) for "synchronous events"? 

in the case where the subscriber makes some model state change rather than a UI element change, how can the caller of `post()` know that all subscribers have finished making their state changes if I use `launch`? 

specifically, what if:

```
@Subscribe fun onSomeModelEvent(event: ModelEvent) /*appropriate coroutine-builder*/ {
  delay(1)
  Model.data = event.data
}


fun complexUpstackBusinessLogic() {
  //....
  runBlocking(Dispatchers.Main){
    //...
    Model.data = null
    eventBus.post(ModelEvent(data = "testing")) 
    println(Model.data) //outputs null or "testing"?
  }
}
```

if `appropriate coroutine-builder` is `launch` its asynchronous. If its `runblocking`, its synchronous.

Work-arounds:
1. I could try to play games with the `Scope` and `Job`s, but I get the feeling it would be hacky. The idea would be that subscribers would mutate some job registry and the poster could then call "wait for recent jobs to finish".
2. I've written an alternative event bus that uses a single coroutine-builder entry-point instead of per-subscriber builders, and supports `@Subscribe suspend fun` methods. This simplfies the problem but it remains. I can use things like `runBlocking(Dispatchers.IO)` and/or `PlatformImpl.enterNestedEventLoop` to skirt dead-lock on that single entry point. 
   - I'll post this attempt at a source-compatible event-bus substitute when I've finished testing it and get my projects gunk off it. 


last point:

maybe all I want is a runtime exception instead of deadlock? Could kotlinx-coroutines (maybe by an optionally implemented method on the dispatchers) speculate about this cercomstance? IE add something like

```kt
if(! privateEventLoop && contextInterceptor.isBlockedByCurrentContext()) throw DeadlockException()

fun ContinuationInterceptor.isBlockedByCurrentContext() = when(this){
  Dispatchers.JavaFx -> Platform.isApplicationThread() 
  Dispatchers.Default -> false //doesnt use thread confinement
  Dispatchers.IO -> false //doesnt use thread confinement
}
```

**Comment by Groostav** [comment]
> `launch(Dispatchers.JavaFx.immediate)` can be used as well

all this does is kick-the-can though, doesn't it? At the first suspension point I'll be right back to where I am now. As per the above example, a `delay(1)` still makes it asynchronous.

**Comment by elizarov** [comment]
@Groostav I have a feeling that you are trying to combine _synchronous_ even processing with _asynchronous_ architecture in some fundamentally non-combinable way. 

> In the case where the subscriber makes some model state change rather than a UI element change, how can the caller of post() know that all subscribers have finished making their state changes if I use launch?

Why would you ever need to do this? Cannot you arhitect your application so that state changes are only _synchronously_ performed by your model classes while UI reacts to those state changes and sends _asynchronous_ events to your models? 

> I've written an alternative event bus that uses a single coroutine-builder entry-point instead of per-subscriber builders, exposes a suspend fun send(event: Any) as an alternative to post(), and supports @Subscribe suspend fun methods. This simplfies the problem but it remains.

It seems to be the fruitful approach. Can you elaborate, please, what the problem remains? 

**Note**, that even if we "fix" deadlock in nested `runBlocking` so that it does not deadlock anymore, then your example with `runBlocking(Main) { delay(1); updateState() }` would only "seem" to work. It will not deadlock, per se, but it will still **block** the main Android thread for the duration of this delay, so I'm not sure if this "fix" (that we, indeed, could do) would be of any help to you.

**Comment by Groostav** [comment]
> @Groostav I have a feeling that you are trying to combine _synchronous_ [event] processing with _asynchronous_ architecture in some fundamentally non-combinable way.

Right, so I hope you can bear with me as I am stumbling through this stuff earnestly. There are a few tools I'll mention below, but my problem was rooted in a misunderstanding of the concept of `runBlocking`.

> Cannot you arhitect your application so that state changes are only _synchronously_ performed by your model classes while UI reacts to those state changes and sends _asynchronous_ events to your models?

We made the choice to mix synchronous business logic with (would-be asynchronous) UI logic as indistinguishable subscribers on a guava event bus. When you post an event, some subscribers must run synchronously, man can run asynchronously. Knowing which is which is hard, _and so I would like to simply make everything ~synchronous~ sequential by default_. 

We then wrote some bad code in java that became ported bad kotlin code. We attempted to handle asynchrony in java with executors and worker threads and we ended up with bugs pertaining to JVM shutdown, worker-pool flooding, etc. 

I want to fix this kotlin code now. I want to use coroutines to do it. 

> It seems to be the fruitful approach. Can you elaborate, please, what the problem remains?
> 
> **Note**, that even if we "fix" deadlock in nested `runBlocking` so that it does not deadlock anymore, then your example with `runBlocking(Main) { delay(1); updateState() }` would only "seem" to work. It will not deadlock, per se, but it will still **block** the main Android thread for the duration of this delay, so I'm not sure if this "fix" (that we, indeed, could do) would be of any help to you.

So: There are 2 things I would like:

1. I would like more visibility into when `runBlocking` will do something silly (IE deadlock). 
2. kotlinx.coroutines should be updated to provide synchronous entry points `runSequentially`

**Regarding 1**: I still believe this should be better than simply deadlocking. To my mind if we can accept the `CoroutineInterceptor` picking up a little bit of speculation, `runBlocking` can reasonably know when it is about to deadlock. Instead of doing that it should throw. I would suggest a kotlin-specific Error (`DispatcherDeadlockError`?) with documentation of an example of the problem.

**Regarding 2**: I'm updating my `fun EventBus.post()` to use [Toolkit.enterNestedEventLoop()](https://gist.github.com/Groostav/49cce5f623d96679abbccfeca295da4a).

I want kotlinx coroutines to offer me a function like 
```kt
fun runSequentiallyInExistingEventLoopOrMakeNewEventLoop(block: CoroutineScope.() -> Unit)
```

This would be how I suggest we move toward that:

```kt
//Builders.kt
fun runSequentially(dispatcher: NestedDispatcher, block: suspend CoroutineScope.() -> Unit){
  //probably you can even re-use `BlockingCoroutine` with a different EventLoop instance
  //you could make this an extension function on CoroutineScope or use a threadlocal to detect if you're down-stack of a `runBlocking` call. 
}

//Dispatchers.kt?
interface NestedDispatcher {
  fun enqueueAndExecute(block: suspend () -> Unit)
}

public sealed class JavaFxDispatcher : MainCoroutineDispatcher(), NestedDispatcher, Delay {
  val toolkit = Toolkit.getToolkit()
  override fun enqueueAndExecute(block: suspend () -> Unit){
    toolkit.doNestedEventLoop { exit -> 
      launch { block(); exit() }
    }
  }
}
```

[JavaFx](https://docs.oracle.com/javase/9/docs/api/javafx/application/Platform.html#enterNestedEventLoop-java.lang.Object-), [Swing](https://docs.oracle.com/javase/8/docs/api/java/awt/EventQueue.html#createSecondaryLoop--), and _android_ ([I think?](https://developer.android.com/reference/android/os/Looper)) support this.

**Comment by elizarov** [comment]
Here is a problem with all those "nested" event loops:
* JavaFx supports it only since JDK 9 (and we target Java 8 now).
* I don't know any public API to create a secondary loop on Android in the same thread.
We can definitely add support for nested loops where they are supported (our internal `EventLoop` architecture was designed with that in mind), but since our main user-base currently comes from Android which does not have the corresponding API, we'd rather focus our efforts on making a better experience for cases where it is not supported.

We've already tried throwing exceptions when `runBlocking` was called at the wrong time in wrong-place and we had to roll-back that change, because people would do things like `runBlockig { doSomethingWeKnowIsFastAndSafe() }`. So, we'd better focus our efforts on deadlock avoidance. So let's keep this issue open for this specific fix.

**Comment by Groostav** [comment]
I've been thinking about a response. But quickly:

>JavaFx supports it only since JDK 9 (and we target Java 8 now).

This is not accurate. I'm on java 8 212 and calling `Toolkit.enterNestedEventLoop` (also exposed as `PlatformImpl.enterNestedEventLoop`, both of which are public but of course get into service loaders and dummy-toolkits and etc. This functionality has been exposde since at least java 8 update 60, probably since at least update 40 (which included some big javafx changes). And I cant find and adaquate javafx version control to really run down exactly where it was introduced. 

I am suprised to learn android does not support a strategy for entering nested event loops.

**Comment by Groostav** [comment]
_I might as well ask for the moon:_

Regarding more precisely what I'm looking for: **what I want is a general purpose way to enter suspension context without blocking**. This is a tall order, but it is precisely what nested event loops, a reasonably common thread-related framework feature, were designed to emulate.

so what I want is:

```kt
fun <R> getOrMakeNestingSuspensionContext(block: suspend () -> R): R ...
```

with:
1. detect re-entrancy. If this is down-stack from a previous `getOrMake...` call, we should simply enqueue `block::resume` on whatever the parent is dispatching on.
   > this is actually what I thought `runBlocking{}` was doing. I was wrong. 
2. using some kind of registry and `Thread.currentThread()`, find out if there is way to enter a nested event loop. Similar to `main`, a service loader could be used to find local frameworks and use the appropriate thread pool. 
3. if none of these strategies yields a loop we can tie into, use the same strategy as runBlocking with a private loop, but publish the resulting loop such that step 1 can find it efficiently.

This would greatly alleviate the red-blue color problem; this would allow authors of new kotlin components that have java/other code up-stack to elegantly use some of the concurrency features of kotlin, rather than forcing authors calling into asynchronous kotlin to continually ask the difficult question "what should my suspension entry point be?" 


> There is one important negative side to this strategy: short of project loom, nested event loops will emulate stack-freezes by appending nested loops "reenter" points to their own stack. This means if you do not have strong bounds on the number of _related_ (nested in the sense of coroutines) `getOrMake` calls, you can hit a stack-overflow. 

For me this would mean subscriber's to a guava event-bus published synchronous event could

```kt
public class MyNastyStatefulController {
  @Subscribe fun handle(domainEvent: DomainEvent) = getOrMakeNestingSuspensionContext {
    model.state += someSuspendingComputation()
    ui.state = updated
  }
}
```

and have that fire synchronously from the caller of `post()` while not blocking. 

**tl;dr** @elizarov please deliver 1x moon to my project.

**Comment by Groostav** [comment]
Alright.

Ultimately I did fork the event bus and I did come up with one that supports `
```kt
@Subscribe suspend fun doStuff(event: Event)
```

I wrote it using `enterSuspensionContext` that combined `Platform.enterNestedEventLoop` with `Dispatchers.JavaFx.immediate` which did the trick. Getting the queue-logic right was difficult and testing it was even more difficult --and I do have some shortcomings there. 

But the end result is bad. 

What were previously performance bugs have become nasty consistency/concurrency bugs. UI elements remain stale (showing the wrong data) while long-running coroutine-subscribers go. More code could be bolted to the problem (a priority queue comes to mind), but I don't have faith that such a solution would work well.

> In hindsight this was obviously going to happen. I had faith that I could magically solve it. I cannot.

At this point I think the best approach is to express subscribers with structured concurrency as extension functions on `CoroutineScope`, which will be bound to the posters CoroutineScope if one is available, or `GlobalScope` if its not.

In this way, a subscriber like this:

```fun
@Subscribe fun doModelTransformOn(event: Event){
  val newData = presumedShortButActuallyExpensiveComputation(event)
}
```

can be upgraded to

```fun
@Subscribe fun CoroutineScope.doModelTransformOn(event: Event){
  launch { //creates a child job to the poster's coroutine scope
    val newData = presumedShortButActuallyExpensiveComputation(event)
  } 
}

and, in exactly the correct fashion, when `suspend fun`s are added down-stack to utility classes, the compiler error will bubble up to this "root" `@Subscribe` method, and the programmer will be asked fairly directly "how do you plan to back-pressure the poster of this event?" and given the tools of structured concurrency to address it. Further, if he does simply `launch` an async child, it will be a child job with a parent that may (and in my case will) have a reasonable `CoroutineName` on it. 
```
